scipy/optimize/minpack.py

```diff
@@ -701,29 +701,25 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         raise ValueError("Method 'lm' only works for unconstrained problems. "
                          "Use 'trf' or 'dogbox' instead.")
 
+    # optimization may produce garbage for float32 inputs, cast them to float64
+
     # NaNs can not be handled
     if check_finite:
-        ydata = np.asarray_chkfinite(ydata)
+        ydata = np.asarray_chkfinite(ydata, float)
     else:
-        ydata = np.asarray(ydata)
+        ydata = np.asarray(ydata, float)
```

Comment:
this code path is never hit by a test

Reply:
But this was already the case before, wasn't it?

Reply:
Yes, my point is that confidence & safety in code changes drops for lines that aren't covered by tests.

Reply:
I'm not really sure how you would go about testing this, apart from trying to fit a dataset with a different dtype and seeing if it failed.

Reply:
I modified another test to exercise that code path.